### DISCUSSION TITLE
RFC-002: Reorganize OpenMLDB modes

### DISCUSSION THREAD

**Discussion by jingchen2222** [4paradigm/OpenMLDB#816] [RFCs]
Please refer to the RFC on Google Drive: [RFC-002: Reorganize OpenMLDB modes](https://docs.google.com/document/d/1PXbymAPocBwqJ6iotEU05seImrWKRZox9D_CT-PDnc4/edit?usp=sharing)

**Status:** Accepted

## Problems
OpenMLDB has many modes which are very confusing. We need to re-organize these modes and provide documents so that the users can easily understand and use them.
In this discussion, we are going to discuss a better way to manage and configure these modes and eliminate unnecessary concepts from users. 
We truly need your suggestions to help us improve our mode configuration system. If you have any questions, suggestions, troubles with our modes and environment usage, please share with us.

### OpenMLDB Service (Approve)

OpenMLDB provides **Online serving** and **Analytical** **training** service.

- **Online serving** is a real-time service that provides feature accessing over one request row in conjunction with its related historical data. 
  - Notice that users should access SQL with a **_request_** input. 
  - For Online serving, SQL queries should be designed extremely performance-sensitive to satisfy the real-time requirements.
  - Users can deploy and access online serving via API SERVER or OpenMLDB SDKs (e.g. JDBC, Python SDK). 
-  **Analytical** **training** provides analytical data batch processing and feature extractions over a set of tables. 
  - Users can analyze data using OepnMLDB SQL. (@jingchen2222: Add link SQL language guide here))
  - Users can perform analytical training via CLI, Spark or using OpenMLDB SDKs (e.g. JDBC SDK, Python SDK).

### Eliminate Performance Sensitive Mode  (Approve)

In v0.3.0, we added performance-sensitive mode into OpenMLDB. This concept is difficult to understand and the mode is hard to manage by users. 

In issue #796, we are going to eliminate **performance-sensitive mode** since the performance-sensitive behaviours can be decided by service type directly. 

- When an engine is running for an online serving service,  it is performance sensitive. 
- When an engine is running for offline training service, it is performance in-sensitive. 

```SQL
-- Assume we have table t1 created in database db1
> USE db1;
-- Analytical training: process data performance-insensitively
> SELECT COUNT(col1) AS col_cnt, SUM(col1) AS sum_col1 from t1; 

-- In order to deploy an Online serving
-- the SQL queries should be designed extremely performance-sensitive to satisfy the real-time requirements.
> DEPLOY demo_deployment SELECT COUNT(col1) OVER w1 AS w1_cnt, SUM(col1) OVER w1 AS w1_sum_col1 FROM t1 WINDOW AS w1 (PARTITION BY col0 ORDER BY std_time ROWS BETWEEN 100 preceding and CURRENT ROW);
```

### Standalone OpenMLDB and Cluster OpenMLDB  (Approve)
There are only two types of OpenMLDB: Standalone OpenMLDB and Cluster OpenMLDB

#### Standalone OpenMLDB  (Approve)
OpenMLDB can be deployed as a standalone DB where the whole data will be stored in one single tablet.

#### Cluster OpenMLDB  (Approve)
OpenMLDB can also be deployed as a distributed DB where data will be distributed in multiple tablets.

**Cluster OpenMLDB**  provides two types of execute_mode: online and offline.

Online execute_mode provides online data manipulation, including `LOAD DATA`, `INSERT`, and **simple data queries**.

Offline execute_mode provides offline data manipulation, including `LOAD DATA` and data queries.

| execute_mode | ManipulateData    | UnSupport Behaviours                                         |
| ------------ | ----------------- | ------------------------------------------------------------ |
| Online       | In Memory         | Complex SELECT Statement, including Group By statement, Having statement, Window statement, aggregation statement, subquery statement, etc |
| Offline      | Local File / HDFS | `INSERT` statement and                                       |

```SQL
-- v0.4.0

> SET @@execute_mode = offline; # using offline storage
> LOAD DATA INFILE ... t1;
> SELECT ... FROM t1;

> DEPLOY SQL; // table data exist, add index
-- ERROR: can't modify index for a table containing data. Please drop the table and...


> SET @@execute_mode = online; # using online storage
> LOAD DATA INFILE ... t1;
> SELECT ... FROM t1;
> SELECT xxxx ; 查看 LIMIT, WHERE
```

### Changes and Additions to Public Interfaces

- eliminate set performance_sensitive = xxx command;

- add `set @@execute_mode=online|offline` in CLI

  

#### v0.4.0

- We will eliminate the **performance-sensitive mode** concept from users. @jingchen2222, **Time estimate:** 5 days, https://github.com/4paradigm/OpenMLDB/issues/796. 
- Docs: guide users to use offline/online in an appropriate way. @denglong, **Time estimate:**, LINK ISSUE:
- Message: throw ERROR message: can't modify index for a table containing data. @denglong **Time estimate:** LINK ISSUE:
- Engine: Support simple SELECT statement under online execute_mode in Cluster OpenMLDB @jingchen2222, **Time estimate: 10d,** https://github.com/4paradigm/OpenMLDB/issues/876
- SQL: support set and show system variable and global variable syntax and logical plan. @jingchen2222 **Time estimate:**  3d, LINK ISSUE:  https://github.com/4paradigm/OpenMLDB/issues/863
- Support `set @@execute_mode=offline|online` in Cluster OpenMLDB's CLI. Disable configuration in Standalone OpenMLDB  @tobe @denglong. **Time estimate:** LINK ISSUE: https://github.com/4paradigm/OpenMLDB/issues/714
- Configure `Cluster_optimize=true` in Cluster OpenMLDB and remove related test. @dl239 Time estimate: LINK ISSUE:
  - Standalone OpenMLDB `Cluster_optimized = false;`
  - Cluster OpenMLDB `Cluster_optimized = true;`
  - Remove `Cluster_optimize=false` in Cluster OpenMLDB and remove related test. 

##### v0.5.0
- Enhancement SQL validate before automatically create indexes. @dl239 ISSUE LINK: 
  - mock performance-insentive compile with requestMode: serving OP: 
    - Request Table path check: DAG structure
    - **GROUP, WHERE, HAVING.** SEVING OP ERROR
  - mock add indexes -- CREATE TABLE INDEX 
  - mock perfomance-sentive compile -- throw ERROR when fails
  - add ts, key indexes
  - create procedure
- Support `TRUNCATE TABLE ...;` SQL syntax @jingchen2222 Link Issues 
- Implement table TRUNCATE: delete all table data. @denglong. Link ISSUE: 

**v1.0.0**
- Support EXPLAIN ONLINE SELECT ... to validate Online serving SQL @jingchen2222 ISSUE LINK: 
- Support New traits in JDBC API, Python SDK, etc

### Future Work

There are many other topics we will discuss in the future.

- Create Procedure standardization
- Eliminate RequetBatch Mode

**Comment by wuyou10206** [comment]
It is recommended to remove the stand-alone mode in the cluster environment

**Comment by zhanghaohit** [comment]
From my understanding, the main difference among the various modes are what the execution engine is used for the queries:
- standalone: openmldb
- cluster: openmldb/spark
- batchMode: openmldb/spark
- requestMode: openmldb
- performance sensitive: openmldb
- performance in-sensitive: openmldb/spark
- Spark mode: spark
- BuildInMemory mode: openmldb

In order to differentiate the different engines for the queries, I think we'd better directly provide an engine or storage option (e.g., spark or openmldb) rather than the various modes and their combinations.

Thus, from the perspectives of the users, there are two storages: offline and online. Offline data is on HDFS or local files, while online data is on openmldb in-memory storage. When they are executing commands, they are aware what kind of storage they are dealing with. Based on the storage types, we can know what to do under the hood.

**Reply by jingchen2222** [reply]
I thought cluster can also be supported by openmldb for request online serving.

**Reply by jingchen2222** [reply]
> Thus, from the perspectives of the users, there are two storages: offline and online. Offline data is on HDFS or local files, while online data is on openmldb in-memory storage. When they are executing commands, they are aware what kind of storage they are dealing with. Based on the storage types, we can know what to do under the hood.

Couldn't agree more. I illustrate the different behaviours in the followed table:


**Standalone**

|              | Online (In-Memory storage) | Offline (HDFS, local files) |
| ------------ | -------------------------- | --------------------------- |
| Request Mode | Performance sensitive. Run by hybridse. | Non-support                 |
| BatchMode    | Performance Insensitive. Run by hybridse. | Non-support                |

**Cluster**

|              | Online (In-Memory storage)              | Offline (HDFS, local files)           |
| ------------ | --------------------------------------- | ------------------------------------- |
| Request Mode | Performance sensitive. Run by hybridse. | Non-support.                          |
| BatchMode    | Non-support                             | Performance Insensitive, Run by spark |

**Reply by jingchen2222** [reply]
So, can we eliminate the spark configuration? Then simply define the behaviours as followed: 
- In standalone environment:
  - Use In-Memory storage
- In cluster environment:
  - Use In-Memory storage for request online serving or SQL deploying
  - Use Offline storage for batch queries.

**Reply by jingchen2222** [reply]
Btw, do uses need to tell the data is online or offline? Will it be transparent for users in the future?

**Reply by lumianph** [reply]
I think online and offline modes, this is the only concept about "mode" that users need to be aware.

**Reply by jingchen2222** [reply]
> I think online and offline modes, this is the only concept about "mode" that users need to be aware.

What does "online" mean? Do you mean online serving with requests? 
Online can be regarded as online storage.

**Reply by lumianph** [reply]
yes, the system currently is running under offline or online mode (online refers to the online serving)

**Reply by zhanghaohit** [reply]
> So, can we eliminate the spark configuration? Then simply define the behaviours as followed:
> 
> * In standalone environment:
>   
>   * Use In-Memory storage
> * In cluster environment:
>   
>   * Use In-Memory storage for request online serving or SQL deploying
>   * Use Offline storage for batch queries.

In cluster environment, users may also want to query the online storage? The normal sql (e.g., select count(*) from t1) is treated as "batch query". So we cannot interpret "cluster" + "batch query" == "offline storage"?

**Reply by jingchen2222** [reply]
So I think we both agree there should only have two modes from the users' perspective.

**Reply by jingchen2222** [reply]
> In my option, the standalone or cluster environment are transparent to the users. What the users have to do is tell us what the storage they want to use, offline or online?

I am concerned about that. In fact, the users don't need to know even the storage details.
What they have to know and care about is if the SQL queried is use for online-serving or offline-training.

**Reply by jingchen2222** [reply]
> In cluster environment, users may also want to query the online storage? The normal sql (e.g., select count(*) from t1) is treated as "batch query". So we cannot interpret "cluster" + "batch query" == "offline storage"?

You're right. So I plan to eliminate the so-called "cluster"+"batch query" mode. It's useless and confusing.

**Reply by lumianph** [reply]
about the cluster and standalone, let us say another word rather than "mode", probably cluster version, or standalone version. this concept is about deployment, different from the execution mode.

**Reply by jingchen2222** [reply]
> about the cluster and standalone, let us say another word rather than "mode", probably cluster version, or standalone version. this concept is about deployment, different from the execution mode.

Agree. And the users don't need to know the difference between them. They are only different in installation.

**Reply by lumianph** [reply]
ideally, either cluster or standalone, the mode concept should always be applicable, and the behavior also should be consistent for both cluster and standalone versions.

**Reply by zhanghaohit** [reply]
> > In my option, the standalone or cluster environment are transparent to the users. What the users have to do is tell us what the storage they want to use, offline or online?
> 
> I am concerned about that. In fact, the users don't need to know even the storage details. What they have to know and care about is if the SQL queried is use for online-serving or offline-training.

I think there are cases where users want to execute batch queries on the "online storage/in-memory storage", say, in the standalone settings.

In these case, "online serving" and "offline serving" are using the same online storage/in-memory storage. They may be confused and they have to create two different tables for "online serving" and "offline serving" although the data in the two tables may be the same.

In addition, the users may also want to explore the data in the "online storage/in-memory storage" using normal sql queries (not request query). These kind of queries are categorised as "online servicing" or "offline serving"?

**Reply by jingchen2222** [reply]
> In addition, the users may also want to explore the data in the "online storage/in-memory storage" using normal sql queries (not request query). These kind of queries are categorised as "online servicing" or "offline serving"?

Why do uses want to query data in online storage? 
Users are going to query data in OpenMLDB DB. Users don't need to know if data is online or offline.

**Reply by jingchen2222** [reply]
> I think there are cases where users want to execute batch queries on the "online storage/in-memory storage", say, in the standalone settings.

I think the "online" concept confuses you @zhanghaohit . That is why I hate these concepts.

Let's say there are only two concepts in OpenMLDB. For instance, request-serving and batch-training.
Based on the two concepts, what kind of troubles you might come across. 

Users load data into OpenMLDB from whatever HDFS or FILE or INSERT statements.
Users query normal SQL with little restrictions.
Users deploy SQL into an online service with strict constraints. Then they can access it with a request. 

I think that's all they need to care about.

**Reply by zhanghaohit** [reply]
> > I think there are cases where users want to execute batch queries on the "online storage/in-memory storage", say, in the standalone settings.
> 
> I think the "online" concept confuses you @zhanghaohit . That is why I hate these concepts.
> 
> Let's say there are only two concepts in OpenMLDB. For instance, request-serving and batch-training. Based on the two concepts, what kind of troubles you might come across.
> 
> Users load data into OpenMLDB from whatever HDFS or FILE or INSERT statements. Users query normal SQL within little restrictions. Users deploy SQL into an online service within strict constraints. Then they can access it with a request.
> 
> I think that's all they need to care about.

Are there two steps for the data loading?
1. load all data to "file storage"
2. load subset of latest data to "in-memory storage"

**Reply by jingchen2222** [reply]
> > > I think there are cases where users want to execute batch queries on the "online storage/in-memory storage", say, in the standalone settings.
> > ...
> 
> Are there two steps for the data loading?
> 
> 1. load all data to "file storage"
> 2. load subset of latest data to "in-memory storage"

I am not sure. It's about data provisioning. @dl239 @tobegit3hub Can you give us more details about it.
However, it's about loading data with one step or two steps. It has nothing to do with execution mode.
So we can simplify the execution mode. We certainly can improve our data loading procedure but that would be another topic we can discuss in the future.

From my point of view, users have to deal with online/offline when they are loading data.

Users unnecessarily know if they are using online data or offline data when they are querying a SQL.

**Comment by jingchen2222** [comment]
From the users' perspective, there should only be two execution modes: RequestMode, BatchMode. We can come up with better names for them if it is necessary. For example, we can use OnlineServing for request mode, OfflineTraining for batchMode
| Execution mode   |  Behaviours              |
| ------------ | -------------------------| 
| Online Serving Mode | Performance sensitive | 
| Training Mode    | Performance Insensitive |

Thus, users only need to take care of SQL performance/index optimization only when they are going to deploy a SQL. Otherwise, users can use our SQL with great tolerance.

From the users' perspective, there are openmldb-cluster version and openmldb-standalone version.
The two versions of DB share the same behaviours most of the time except that the installation and deployment ways are different,

**Reply by jingchen2222** [reply]
In order to achieve the above design purpose, we have to re-define behaviours:
1. Eliminate **BuiltInMemery** Mode. Standalone runs on memory automatically.
2. Eliminate **Spark** Mode. Cluster offline training runs on spark automatically.

### OpenMLDB standalone
Online serving and offline training both run in memory automatically.

### OpenMLDB cluster
Online serving runs in memory automatically.
Offline training runs in spark automatically, although users don't need to know that.

**Comment by jingchen2222** [comment]
@zhanghaohit We should do something to load online data and load offline data without using `mode=BuildInMemeory` or `mode=spark`. Ideally, we don't want to add a mode for that. We should discuss the topic next week.
